In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [4]:
telecom = pd.read_csv("telecom_churn_data.csv")
telecom.head

<bound method NDFrame.head of        mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
0         7000842753        109             0.0             0.0   
1         7001865778        109             0.0             0.0   
2         7001625959        109             0.0             0.0   
3         7001204172        109             0.0             0.0   
4         7000142493        109             0.0             0.0   
5         7000286308        109             0.0             0.0   
6         7001051193        109             0.0             0.0   
7         7000701601        109             0.0             0.0   
8         7001524846        109             0.0             0.0   
9         7001864400        109             0.0             0.0   
10        7000471350        109             0.0             0.0   
11        7002006969        109             0.0             0.0   
12        7001419799        109             0.0             0.0   
13        7002191713        109 

In [5]:
# Let's check the dimensions of the dataframe
telecom.shape

(99999, 226)

In [6]:
# let's look at the statistical view of the dataframe
telecom.describe()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count   9.999900e+04    99999.0         98981.0         98981.0   
mean    7.001207e+09      109.0             0.0             0.0   
std     6.956694e+05        0.0             0.0             0.0   
min     7.000000e+09      109.0             0.0             0.0   
25%     7.000606e+09      109.0             0.0             0.0   
50%     7.001205e+09      109.0             0.0             0.0   
75%     7.001812e+09      109.0             0.0             0.0   
max     7.002411e+09      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8        arpu_9  \
count         98981.0  99999.000000  99999.000000  99999.000000  99999.000000   
mean              0.0    282.987358    278.536648    279.154731    261.645069   
std               0.0    328.439770    338.156291    344.474791    341.998630   
min               0.0  -2258.709000  -2014.045000   -945.808000  -1899.505000   
25%               0.0     93.411500     86.980500     84.126000     62.685000   
50%               0.0    197.704000    191.640000    192.080000    176.849000   
75%               0.0    371.060000    365.344500    369.370500    353.466500   
max               0.0  27731.088000  35145.834000  33543.624000  38805.617000   

        onnet_mou_6      ...        sachet_3g_9     fb_user_6     fb_user_7  \
count  96062.000000      ...       99999.000000  25153.000000  25571.000000   
mean     132.395875      ...           0.084581      0.914404      0.908764   
std      297.207406      ...           0.650457      0.279772      0.287950   
min        0.000000      ...           0.000000      0.000000      0.000000   
25%        7.380000      ...           0.000000      1.000000      1.000000   
50%       34.310000      ...           0.000000      1.000000      1.000000   
75%      118.740000      ...           0.000000      1.000000      1.000000   
max     7376.710000      ...          49.000000      1.000000      1.000000   

          fb_user_8     fb_user_9           aon    aug_vbc_3g    jul_vbc_3g  \
count  26339.000000  25922.000000  99999.000000  99999.000000  99999.000000   
mean       0.890808      0.860968   1219.854749     68.170248     66.839062   
std        0.311885      0.345987    954.733842    267.580450    271.201856   
min        0.000000      0.000000    180.000000      0.000000      0.000000   
25%        1.000000      1.000000    467.000000      0.000000      0.000000   
50%        1.000000      1.000000    863.000000      0.000000      0.000000   
75%        1.000000      1.000000   1807.500000      0.000000      0.000000   
max        1.000000      1.000000   4337.000000  12916.220000   9165.600000   

         jun_vbc_3g    sep_vbc_3g  
count  99999.000000  99999.000000  
mean      60.021204      3.299373  
std      253.938223     32.408353  
min        0.000000      0.000000  
25%        0.000000      0.000000  
50%        0.000000      0.000000  
75%        0.000000      0.000000  
max    11166.210000   2618.570000  

[8 rows x 214 columns]

In [7]:
telecom.info(verbose=True,null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 226 columns):
mobile_number               99999 non-null int64
circle_id                   99999 non-null int64
loc_og_t2o_mou              98981 non-null float64
std_og_t2o_mou              98981 non-null float64
loc_ic_t2o_mou              98981 non-null float64
last_date_of_month_6        99999 non-null object
last_date_of_month_7        99398 non-null object
last_date_of_month_8        98899 non-null object
last_date_of_month_9        98340 non-null object
arpu_6                      99999 non-null float64
arpu_7                      99999 non-null float64
arpu_8                      99999 non-null float64
arpu_9                      99999 non-null float64
onnet_mou_6                 96062 non-null float64
onnet_mou_7                 96140 non-null float64
onnet_mou_8                 94621 non-null float64
onnet_mou_9                 92254 non-null float64
offnet_mou_6                960

In [8]:
#convert 'last_date_of_month_6','date_of_last_rech_6','date_of_last_rech_data_6 timestamp' columns to datetime

telecom['last_date_of_month_6'] = telecom['last_date_of_month_6'].str.replace("-", "/")
telecom['last_date_of_month_6'] = pd.to_datetime(telecom['last_date_of_month_6'], dayfirst=True)
telecom['last_date_of_month_7'] = telecom['last_date_of_month_7'].str.replace("-", "/")
telecom['last_date_of_month_7'] = pd.to_datetime(telecom['last_date_of_month_7'], dayfirst=True)
telecom['last_date_of_month_8'] = telecom['last_date_of_month_8'].str.replace("-", "/")
telecom['last_date_of_month_8'] = pd.to_datetime(telecom['last_date_of_month_8'], dayfirst=True)
telecom['last_date_of_month_9'] = telecom['last_date_of_month_9'].str.replace("-", "/")
telecom['last_date_of_month_9'] = pd.to_datetime(telecom['last_date_of_month_9'], dayfirst=True)

In [9]:
telecom['date_of_last_rech_6'] = telecom['date_of_last_rech_6'].str.replace("-", "/")
telecom['date_of_last_rech_6'] = pd.to_datetime(telecom['date_of_last_rech_6'], dayfirst=True)
telecom['date_of_last_rech_7'] = telecom['date_of_last_rech_7'].str.replace("-", "/")
telecom['date_of_last_rech_7'] = pd.to_datetime(telecom['date_of_last_rech_7'], dayfirst=True)
telecom['date_of_last_rech_8'] = telecom['date_of_last_rech_8'].str.replace("-", "/")
telecom['date_of_last_rech_8'] = pd.to_datetime(telecom['date_of_last_rech_8'], dayfirst=True)
telecom['date_of_last_rech_9'] = telecom['date_of_last_rech_9'].str.replace("-", "/")
telecom['date_of_last_rech_9'] = pd.to_datetime(telecom['date_of_last_rech_9'], dayfirst=True)

In [10]:
telecom.head(10)

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   
5     7000286308        109             0.0             0.0             0.0   
6     7001051193        109             0.0             0.0             0.0   
7     7000701601        109             0.0             0.0             0.0   
8     7001524846        109             0.0             0.0             0.0   
9     7001864400        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0           2014-06-30           2014-07-31           2014-08-31   
1           2014-06-30           2014-07-31           2014-08-31   
2           2014-06-30           2014-07-31           2014-08-31   
3           2014-06-30           2014-07-31           2014-08-31   
4           2014-06-30           2014-07-31           2014-08-31   
5           2014-06-30           2014-07-31           2014-08-31   
6           2014-06-30           2014-07-31           2014-08-31   
7           2014-06-30           2014-07-31           2014-08-31   
8           2014-06-30           2014-07-31           2014-08-31   
9           2014-06-30           2014-07-31           2014-08-31   

  last_date_of_month_9    arpu_6     ...      sachet_3g_9  fb_user_6  \
0           2014-09-30   197.385     ...                0        1.0   
1           2014-09-30    34.047     ...                0        NaN   
2           2014-09-30   167.690     ...                0        NaN   
3           2014-09-30   221.338     ...                0        NaN   
4           2014-09-30   261.636     ...                0        0.0   
5           2014-09-30    50.258     ...                0        NaN   
6           2014-09-30   429.023     ...                0        NaN   
7           2014-09-30  1069.180     ...                0        NaN   
8           2014-09-30   378.721     ...                0        NaN   
9           2014-09-30   119.518     ...                0        NaN   

   fb_user_7  fb_user_8  fb_user_9   aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  \
0        1.0        1.0        NaN   968       30.40        0.00      101.20   
1        1.0        1.0        NaN  1006        0.00        0.00        0.00   
2        NaN        NaN        1.0  1103        0.00        0.00        4.17   
3        NaN        NaN        NaN  2491        0.00        0.00        0.00   
4        NaN        NaN        NaN  1526        0.00        0.00        0.00   
5        NaN        NaN        NaN  1471        0.00        0.00        0.00   
6        NaN        NaN        NaN  1673        0.00        0.00        0.00   
7        NaN        NaN        NaN   802       57.74       19.38       18.74   
8        1.0        1.0        NaN   315       21.03      910.65      122.16   
9        1.0        NaN        NaN   902        0.00        0.00        0.00   

   sep_vbc_3g  
0        3.58  
1        0.00  
2        0.00  
3        0.00  
4        0.00  
5        0.00  
6        0.00  
7        0.00  
8        0.00  
9        0.00  

[10 rows x 226 columns]

In [11]:
#lets make another copy of the original dataset so that it may come handy later on during modelling
telecom2 = telecom.copy()

In [12]:
#lets make seperate groups for column names of date, categpory, numeric vallues & ID columns, so that we can segregate it properly

date = ['last_date_of_month_6',
             'last_date_of_month_7',
             'last_date_of_month_8',
             'last_date_of_month_9',
             'date_of_last_rech_6',
             'date_of_last_rech_7',
             'date_of_last_rech_8',
             'date_of_last_rech_9',
             'date_of_last_rech_data_6',
             'date_of_last_rech_data_7',
             'date_of_last_rech_data_8',
             'date_of_last_rech_data_9'
            ]

catg =  ['night_pck_user_6',
             'night_pck_user_7',
             'night_pck_user_8',
             'night_pck_user_9',
             'fb_user_6',
             'fb_user_7',
             'fb_user_8',
             'fb_user_9'
            ]
ID = ['circle_id','mobile_number']

In [13]:
number = [col for col in telecom.columns if col not in date + catg + ID]

In [14]:
# let us first extract list of columns containing recharge amount
recharge_amnt_cols =  telecom.columns[telecom.columns.str.contains('rech_amt|rech_data')]

print(recharge_amnt_cols)

Index(['total_rech_amt_6', 'total_rech_amt_7', 'total_rech_amt_8',
       'total_rech_amt_9', 'max_rech_amt_6', 'max_rech_amt_7',
       'max_rech_amt_8', 'max_rech_amt_9', 'date_of_last_rech_data_6',
       'date_of_last_rech_data_7', 'date_of_last_rech_data_8',
       'date_of_last_rech_data_9', 'total_rech_data_6', 'total_rech_data_7',
       'total_rech_data_8', 'total_rech_data_9', 'max_rech_data_6',
       'max_rech_data_7', 'max_rech_data_8', 'max_rech_data_9',
       'av_rech_amt_data_6', 'av_rech_amt_data_7', 'av_rech_amt_data_8',
       'av_rech_amt_data_9'],
      dtype='object')


In [15]:
# look at missing value ratio in each column
telecom.isnull().sum()*100/telecom.shape[0]

mobile_number            0.000000
circle_id                0.000000
loc_og_t2o_mou           1.018010
std_og_t2o_mou           1.018010
loc_ic_t2o_mou           1.018010
last_date_of_month_6     0.000000
last_date_of_month_7     0.601006
last_date_of_month_8     1.100011
last_date_of_month_9     1.659017
arpu_6                   0.000000
arpu_7                   0.000000
arpu_8                   0.000000
arpu_9                   0.000000
onnet_mou_6              3.937039
onnet_mou_7              3.859039
onnet_mou_8              5.378054
onnet_mou_9              7.745077
offnet_mou_6             3.937039
offnet_mou_7             3.859039
offnet_mou_8             5.378054
offnet_mou_9             7.745077
roam_ic_mou_6            3.937039
roam_ic_mou_7            3.859039
roam_ic_mou_8            5.378054
roam_ic_mou_9            7.745077
roam_og_mou_6            3.937039
roam_og_mou_7            3.859039
roam_og_mou_8            5.378054
roam_og_mou_9            7.745077
loc_og_t2t_mou

we observe that approx 74-75% values for average recharge amount, max recharge data and total recharge date and date of data recharge for mobile internet data is missing.

In [16]:
#we again see some recharge columns having minimum value of 1 but also there are some which doesnt
recharge = ['total_rech_data_6', 'total_rech_data_7', 'total_rech_data_8', 'total_rech_data_9',
                 'count_rech_2g_6', 'count_rech_2g_7', 'count_rech_2g_8', 'count_rech_2g_9',
                 'count_rech_3g_6', 'count_rech_3g_7', 'count_rech_3g_8', 'count_rech_3g_9',
                 'max_rech_data_6', 'max_rech_data_7', 'max_rech_data_8', 'max_rech_data_9',
                 'av_rech_amt_data_6', 'av_rech_amt_data_7', 'av_rech_amt_data_8', 'av_rech_amt_data_9',
                 ]

telecom[recharge].describe(include='all')

total_rech_data_6  total_rech_data_7  total_rech_data_8  \
count       25153.000000       25571.000000       26339.000000   
mean            2.463802           2.666419           2.651999   
std             2.789128           3.031593           3.074987   
min             1.000000           1.000000           1.000000   
25%             1.000000           1.000000           1.000000   
50%             1.000000           1.000000           1.000000   
75%             3.000000           3.000000           3.000000   
max            61.000000          54.000000          60.000000   

       total_rech_data_9  count_rech_2g_6  count_rech_2g_7  count_rech_2g_8  \
count       25922.000000     25153.000000     25571.000000     26339.000000   
mean            2.441170         1.864668         2.044699         2.016288   
std             2.516339         2.570254         2.768332         2.720132   
min             1.000000         0.000000         0.000000         0.000000   
25%             1.000000         1.000000         1.000000         1.000000   
50%             2.000000         1.000000         1.000000         1.000000   
75%             3.000000         2.000000         2.000000         2.000000   
max            84.000000        42.000000        48.000000        44.000000   

       count_rech_2g_9  count_rech_3g_6  count_rech_3g_7  count_rech_3g_8  \
count     25922.000000     25153.000000     25571.000000     26339.000000   
mean          1.781807         0.599133         0.621720         0.635711   
std           2.214701         1.274428         1.394524         1.422827   
min           0.000000         0.000000         0.000000         0.000000   
25%           1.000000         0.000000         0.000000         0.000000   
50%           1.000000         0.000000         0.000000         0.000000   
75%           2.000000         1.000000         1.000000         1.000000   
max          40.000000        29.000000        35.000000        45.000000   

       count_rech_3g_9  max_rech_data_6  max_rech_data_7  max_rech_data_8  \
count     25922.000000     25153.000000     25571.000000     26339.000000   
mean          0.659363       126.393392       126.729459       125.717301   
std           1.411513       108.477235       109.765267       109.437851   
min           0.000000         1.000000         1.000000         1.000000   
25%           0.000000        25.000000        25.000000        25.000000   
50%           0.000000       145.000000       145.000000       145.000000   
75%           1.000000       177.000000       177.000000       179.000000   
max          49.000000      1555.000000      1555.000000      1555.000000   

       max_rech_data_9  av_rech_amt_data_6  av_rech_amt_data_7  \
count      25922.00000        25153.000000        25571.000000   
mean         124.94144          192.600982          200.981292   
std          111.36376          192.646318          196.791224   
min            1.00000            1.000000            0.500000   
25%           25.00000           82.000000           92.000000   
50%          145.00000          154.000000          154.000000   
75%          179.00000          252.000000          252.000000   
max         1555.00000         7546.000000         4365.000000   

       av_rech_amt_data_8  av_rech_amt_data_9  
count        26339.000000        25922.000000  
mean           197.526489          192.734315  
std            191.301305          188.400286  
min              0.500000            1.000000  
25%             87.000000           69.000000  
50%            154.000000          164.000000  
75%            252.000000          252.000000  
max           4076.000000         4061.000000

We see from above the values are consistent with each other, also as the minimum value is 1 we can impute the NA values by 0, meaning that there was no recharge done by the customer.

### We put zeros in null recharge amounts

In [17]:
#first lets have a view of customers who didn't recharge any
telecom.loc[telecom.total_rech_data_6.isnull() & telecom.date_of_last_rech_data_6.isnull(), ["total_rech_data_6", "date_of_last_rech_data_6"]].head(10)


total_rech_data_6 date_of_last_rech_data_6
1                 NaN                      NaN
2                 NaN                      NaN
3                 NaN                      NaN
5                 NaN                      NaN
6                 NaN                      NaN
7                 NaN                      NaN
8                 NaN                      NaN
9                 NaN                      NaN
10                NaN                      NaN
11                NaN                      NaN

In [18]:
#we fill the null recharge value with an equivalent zero since it will help us analyse properly
zero_recharge = ['total_rech_data_6', 'total_rech_data_7', 'total_rech_data_8', 'total_rech_data_9',
        'av_rech_amt_data_6', 'av_rech_amt_data_7', 'av_rech_amt_data_8', 'av_rech_amt_data_9',
        'max_rech_data_6', 'max_rech_data_7', 'max_rech_data_8', 'max_rech_data_9'
       ]

telecom[zero_recharge] = telecom[zero_recharge].apply(lambda x: x.fillna(0))

In [19]:
#lets hav a view of the missing values are properly imputed or not
telecom[zero_recharge].isnull().sum()*100/telecom.shape[1]

total_rech_data_6     0.0
total_rech_data_7     0.0
total_rech_data_8     0.0
total_rech_data_9     0.0
av_rech_amt_data_6    0.0
av_rech_amt_data_7    0.0
av_rech_amt_data_8    0.0
av_rech_amt_data_9    0.0
max_rech_data_6       0.0
max_rech_data_7       0.0
max_rech_data_8       0.0
max_rech_data_9       0.0
dtype: float64

In [ ]:
#A good look of the zero rcharge
telecom[zero_recharge].describe(include='all')

total_rech_data_6  total_rech_data_7  total_rech_data_8  \
count       99999.000000       99999.000000       99999.000000   
mean            0.619726           0.681837           0.698517   
std             1.760541           1.924382           1.963417   
min             0.000000           0.000000           0.000000   
25%             0.000000           0.000000           0.000000   
50%             0.000000           0.000000           0.000000   
75%             1.000000           1.000000           1.000000   
max            61.000000          54.000000          60.000000   

       total_rech_data_9  av_rech_amt_data_6  av_rech_amt_data_7  \
count       99999.000000        99999.000000        99999.000000   
mean            0.632806           48.445409           51.393440   
std             1.669040          127.743863          132.629365   
min             0.000000            0.000000            0.000000   
25%             0.000000            0.000000            0.000000   
50%             0.000000            0.000000            0.000000   
75%             1.000000            8.250000           17.000000   
max            84.000000         7546.000000         4365.000000   

       av_rech_amt_data_8  av_rech_amt_data_9  max_rech_data_6  \
count        99999.000000        99999.000000     99999.000000   
mean            52.027022           49.961089        31.792048   
std            131.182609          127.804280        77.248778   
min              0.000000            0.000000         0.000000   
25%              0.000000            0.000000         0.000000   
50%              0.000000            0.000000         0.000000   
75%             23.000000           17.000000         8.000000   
max           4076.000000         4061.000000      1555.000000   

       max_rech_data_7  max_rech_data_8  max_rech_data_9  
count     99999.000000     99999.000000     99999.000000  
mean         32.406314        33.113011        32.387644  
std          78.342435        78.872739        78.818696  
min           0.000000         0.000000         0.000000  
25%           0.000000         0.000000         0.000000  
50%           0.000000         0.000000         0.000000  
75%          14.000000        17.000000        17.000000  
max        1555.000000      1555.000000      1555.000000

In [ ]:
#now we drop Id and date columns
telecom = telecom.drop(ID + date, axis=1)
telecom.shape
#we see the columns have dropped from 226 to 212

(99999, 212)

### Now we put -1 in place of missing categorical values 

In [ ]:
#now lets replace missing values in category column with '-1' 
telecom[catg] = telecom[catg].apply(lambda x: x.fillna(-1))
telecom[catg].isnull().sum()*100/telecom.shape[0]

night_pck_user_6    0.0
night_pck_user_7    0.0
night_pck_user_8    0.0
night_pck_user_9    0.0
fb_user_6           0.0
fb_user_7           0.0
fb_user_8           0.0
fb_user_9           0.0
dtype: float64

### dropping suitable columns of null values more than 70% 

In [ ]:
null_max = 0.7
original_col = telecom.shape[1]
suitable_cols = list(telecom.apply(lambda column: True if column.isnull().sum()/telecom.shape[0] < null_max else False))
drop_null_values = pd.DataFrame({'feat':telecom.columns , 'suitable': suitable_cols})
drop_null_values.loc[drop_null_values.suitable == True,:]

feat  suitable
0      loc_og_t2o_mou      True
1      std_og_t2o_mou      True
2      loc_ic_t2o_mou      True
3              arpu_6      True
4              arpu_7      True
5              arpu_8      True
6              arpu_9      True
7         onnet_mou_6      True
8         onnet_mou_7      True
9         onnet_mou_8      True
10        onnet_mou_9      True
11       offnet_mou_6      True
12       offnet_mou_7      True
13       offnet_mou_8      True
14       offnet_mou_9      True
15      roam_ic_mou_6      True
16      roam_ic_mou_7      True
17      roam_ic_mou_8      True
18      roam_ic_mou_9      True
19      roam_og_mou_6      True
20      roam_og_mou_7      True
21      roam_og_mou_8      True
22      roam_og_mou_9      True
23   loc_og_t2t_mou_6      True
24   loc_og_t2t_mou_7      True
25   loc_og_t2t_mou_8      True
26   loc_og_t2t_mou_9      True
27   loc_og_t2m_mou_6      True
28   loc_og_t2m_mou_7      True
29   loc_og_t2m_mou_8      True
..                ...       ...
174       vol_3g_mb_9      True
183  night_pck_user_6      True
184  night_pck_user_7      True
185  night_pck_user_8      True
186  night_pck_user_9      True
187      monthly_2g_6      True
188      monthly_2g_7      True
189      monthly_2g_8      True
190      monthly_2g_9      True
191       sachet_2g_6      True
192       sachet_2g_7      True
193       sachet_2g_8      True
194       sachet_2g_9      True
195      monthly_3g_6      True
196      monthly_3g_7      True
197      monthly_3g_8      True
198      monthly_3g_9      True
199       sachet_3g_6      True
200       sachet_3g_7      True
201       sachet_3g_8      True
202       sachet_3g_9      True
203         fb_user_6      True
204         fb_user_7      True
205         fb_user_8      True
206         fb_user_9      True
207               aon      True
208        aug_vbc_3g      True
209        jul_vbc_3g      True
210        jun_vbc_3g      True
211        sep_vbc_3g      True

[196 rows x 2 columns]

In [ ]:
#After creating suitable missing value drop columns, we drop the suitable columns
telecom = telecom.loc[:, suitable_cols]
dropped = telecom.shape[1] - original_col
print("{0} columns dropped.".format(dropped))

-16 columns dropped.


### Let's filter out the customers with high value churn 

#### total data & with (data+talktime) recharge amount 

In [ ]:
#now lets compute the total recharging data amount for June and July
telecom['total_data_rech_6'] = telecom.av_rech_amt_data_6 * telecom.total_rech_data_6
telecom['total_data_rech_7'] = telecom.av_rech_amt_data_7 * telecom.total_rech_data_7
#now lets compute talktime + data recharging amount for June and July
telecom['amt_data_6'] =  telecom.total_data_rech_6 + telecom.total_rech_amt_6
telecom['amt_data_7'] =  telecom.total_data_rech_7 + telecom.total_rech_amt_7

#### average recharge amount with 70th percentile amount

In [ ]:
#average recharge amount by customer in June and July
telecom['avg_amt_data_6_7'] = (telecom.amt_data_6 + telecom.amt_data_7)/2
# look at the 70th percentile recharge amount
print("70th percentile recharge amount: {0}".format(telecom.avg_amt_data_6_7.quantile(0.7)))

70th percentile recharge amount: 478.0


In [ ]:
#display only those customers who have recharge amount is greater than 70th percentile amount
telecom_filtered = telecom.loc[telecom.avg_amt_data_6_7 >= telecom.avg_amt_data_6_7.quantile(0.7), :]
telecom_filtered = telecom_filtered.reset_index(drop=True)
telecom_filtered.shape

(30001, 201)

In [ ]:
# we can now drop the extra columns created to filter high-value customers
telecom_filtered = telecom_filtered.drop(['total_data_rech_6', 'total_data_rech_7',
                                      'amt_data_6', 'amt_data_7', 'avg_amt_data_6_7'], axis=1)
telecom_filtered.shape

(30001, 196)

### creating variables for churned customers.

In [ ]:
#total incoming and outgoing rates 
telecom_filtered['total_calls_mou_9'] = telecom_filtered.total_og_mou_9 + telecom_filtered.total_ic_mou_9
#different internet speed of data consumed
telecom_filtered['total_internet_mb_9'] = telecom_filtered.vol_3g_mb_9 + telecom_filtered.vol_2g_mb_9
#lets denote it by 1 for churn and 0 for not churn
telecom_filtered['churn'] = telecom_filtered.apply(lambda row: 1 if (row.total_calls_mou_9 == 0 and row.total_internet_mb_9 == 0) else 0, axis=1)
#drop extra variables created
telecom_filtered = telecom_filtered.drop(['total_calls_mou_9', 'total_internet_mb_9'], axis=1)
#we change the datatypes to category to find out the churned customers
telecom_filtered.churn = telecom_filtered.churn.astype("category")

In [ ]:
# let us check what's the % of churned customers
churn_value_counts  = telecom_filtered.churn.value_counts()
print("Total_Churn_Customers     = ",churn_value_counts[1])
print("Total_Non_Churn_Customers = ",churn_value_counts[0])
print("Churn %")
print(telecom_filtered.churn.value_counts()*100/telecom_filtered.shape[0])

Total_Churn_Customers     =  2441
Total_Non_Churn_Customers =  27560
Churn %
0    91.863605
1     8.136395
Name: churn, dtype: float64


We see that approx. 8.14% of high value customers have churned

### now we create few extra features

In [ ]:
#here we calculate the difference between the average recharge amount of 6th and 7th month and the recharge amount of 8th amount
telecom_filtered['roam_og_mou_diff'] = telecom_filtered.roam_og_mou_8 - ((telecom_filtered.roam_og_mou_6 + telecom_filtered.roam_og_mou_7)/2)

telecom_filtered['loc_og_mou_diff'] = telecom_filtered.loc_og_mou_8 - ((telecom_filtered.loc_og_mou_6 + telecom_filtered.loc_og_mou_7)/2)

telecom_filtered['std_og_mou_diff'] = telecom_filtered.std_og_mou_8 - ((telecom_filtered.std_og_mou_6 + telecom_filtered.std_og_mou_7)/2)

telecom_filtered['isd_og_mou_diff'] = telecom_filtered.isd_og_mou_8 - ((telecom_filtered.isd_og_mou_6 + telecom_filtered.isd_og_mou_7)/2)

telecom_filtered['spl_og_mou_diff'] = telecom_filtered.spl_og_mou_8 - ((telecom_filtered.spl_og_mou_6 + telecom_filtered.spl_og_mou_7)/2)

telecom_filtered['total_og_mou_diff'] = telecom_filtered.total_og_mou_8 - ((telecom_filtered.total_og_mou_6 + telecom_filtered.total_og_mou_7)/2)

telecom_filtered['loc_ic_mou_diff'] = telecom_filtered.loc_ic_mou_8 - ((telecom_filtered.loc_ic_mou_6 + telecom_filtered.loc_ic_mou_7)/2)


In [ ]:
telecom_filtered['std_ic_mou_diff'] = telecom_filtered.std_ic_mou_8 - ((telecom_filtered.std_ic_mou_6 + telecom_filtered.std_ic_mou_7)/2)

telecom_filtered['isd_ic_mou_diff'] = telecom_filtered.isd_ic_mou_8 - ((telecom_filtered.isd_ic_mou_6 + telecom_filtered.isd_ic_mou_7)/2)

telecom_filtered['spl_ic_mou_diff'] = telecom_filtered.spl_ic_mou_8 - ((telecom_filtered.spl_ic_mou_6 + telecom_filtered.spl_ic_mou_7)/2)

telecom_filtered['total_ic_mou_diff'] = telecom_filtered.total_ic_mou_8 - ((telecom_filtered.total_ic_mou_6 + telecom_filtered.total_ic_mou_7)/2)

telecom_filtered['total_rech_num_diff'] = telecom_filtered.total_rech_num_8 - ((telecom_filtered.total_rech_num_6 + telecom_filtered.total_rech_num_7)/2)

telecom_filtered['total_rech_amt_diff'] = telecom_filtered.total_rech_amt_8 - ((telecom_filtered.total_rech_amt_6 + telecom_filtered.total_rech_amt_7)/2)

telecom_filtered['max_rech_amt_diff'] = telecom_filtered.max_rech_amt_8 - ((telecom_filtered.max_rech_amt_6 + telecom_filtered.max_rech_amt_7)/2)

telecom_filtered['total_rech_data_diff'] = telecom_filtered.total_rech_data_8 - ((telecom_filtered.total_rech_data_6 + telecom_filtered.total_rech_data_7)/2)


In [ ]:
telecom_filtered['arpu_diff'] = telecom_filtered.arpu_8 - ((telecom_filtered.arpu_6 + telecom_filtered.arpu_7)/2)

telecom_filtered['onnet_mou_diff'] = telecom_filtered.onnet_mou_8 - ((telecom_filtered.onnet_mou_6 + telecom_filtered.onnet_mou_7)/2)

telecom_filtered['offnet_mou_diff'] = telecom_filtered.offnet_mou_8 - ((telecom_filtered.offnet_mou_6 + telecom_filtered.offnet_mou_7)/2)

telecom_filtered['roam_ic_mou_diff'] = telecom_filtered.roam_ic_mou_8 - ((telecom_filtered.roam_ic_mou_6 + telecom_filtered.roam_ic_mou_7)/2)

telecom_filtered['max_rech_data_diff'] = telecom_filtered.max_rech_data_8 - ((telecom_filtered.max_rech_data_6 + telecom_filtered.max_rech_data_7)/2)

telecom_filtered['av_rech_amt_data_diff'] = telecom_filtered.av_rech_amt_data_8 - ((telecom_filtered.av_rech_amt_data_6 + telecom_filtered.av_rech_amt_data_7)/2)

telecom_filtered['vol_2g_mb_diff'] = telecom_filtered.vol_2g_mb_8 - ((telecom_filtered.vol_2g_mb_6 + telecom_filtered.vol_2g_mb_7)/2)

telecom_filtered['vol_3g_mb_diff'] = telecom_filtered.vol_3g_mb_8 - ((telecom_filtered.vol_3g_mb_6 + telecom_filtered.vol_3g_mb_7)/2)


In [ ]:
# let's extract all list of 9th Month (churned month) Columns using regex
month_9 = telecom.filter(regex='9$', axis=1).columns
len(month_9)

47

In [ ]:
# drop all columns corresponding to the attributes of churn phase of 9th month
month_9 = telecom.filter(regex='9$', axis=1).columns

In [ ]:
#lets check category and number columns have any values that belong to month_9
catg = [col for col in catg if col not in month_9]
catg.append('churn')
number = [col for col in telecom_filtered.columns if col not in catg]

In [ ]:
#now we update the colum data types
telecom_filtered[number] = telecom_filtered[number].apply(pd.to_numeric)
telecom_filtered[catg] = telecom_filtered[catg].apply(lambda column: column.astype("category"), axis=0)

#### Outlier treatment

In [ ]:
#defining a function for treating outliers via k capping technique (source: STACK OVERFLOW)
def outliers(array, k=3):
    u_lim = array.mean() + k*array.std()
    l_lim = array.mean() - k*array.std()
    array[array>u_lim] = u_lim
    array[array<l_lim] = l_lim
    return array

In [ ]:
# cap outliers in the numeric columns
telecom_filtered[number] = telecom_filtered[number].apply(outliers, axis=0)

### Dropping missing values in filtered telecom df

In [ ]:
round(100*(telecom_filtered.isnull().sum()/len(telecom_filtered.index)), 2)

In [ ]:
#creating another filter table impute missing values in the rest of the columns
filtered_telecom_df = telecom_filtered.dropna(how='any')

In [ ]:
round(100*(filtered_telecom_df.isnull().sum()/len(filtered_telecom_df)), 2)

### Modelling

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import FeatureUnion
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.svm import SVC
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import roc_auc_score
from imblearn.metrics import sensitivity_specificity_support
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score

In [ ]:
#conversion of churn values to numeric datatype
filtered_telecom_df['churn'] = pd.to_numeric(filtered_telecom_df['churn'])
#train and test split
X = filtered_telecom_df.drop("churn", axis = 1)
y = filtered_telecom_df.churn
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, stratify = y, random_state = 4)

In [ ]:
#train and test sets
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

In [ ]:
#grouping the categorical value
train = pd.concat([X_train, y_train], axis=1)
train.groupby('night_pck_user_6').churn.mean()
train.groupby('night_pck_user_7').churn.mean()
train.groupby('night_pck_user_8').churn.mean()
train.groupby('fb_user_6').churn.mean()
train.groupby('fb_user_7').churn.mean()
train.groupby('fb_user_8').churn.mean()

In [ ]:
#group by values replaces with categorical values
mapp = {'night_pck_user_6' : {-1: 0.099165, 0: 0.066797, 1: 0.087838},
           'night_pck_user_7' : {-1: 0.115746, 0: 0.055494, 1: 0.051282},
           'night_pck_user_8' : {-1: 0.141108, 0: 0.029023, 1: 0.016194},
           'fb_user_6'        : {-1: 0.099165, 0: 0.069460, 1: 0.067124},
           'fb_user_7'        : {-1: 0.115746, 0: 0.059305, 1: 0.055082},
           'fb_user_8'        : {-1: 0.141108, 0: 0.066887, 1: 0.024463}
          }
X_train.replace(mapp, inplace = True)
X_test.replace(mapp, inplace = True)

In [ ]:
# check data type of categorical columns - make sure they are numeric
X_train[[col for col in catg if col not in ['churn']]].info()

In [ ]:
#we scale the train data and applying pca 
pca = Pipeline([('scaler', StandardScaler()), ('pca', PCA())])
pca.fit(X_train)
telecom_pca = pca.fit_transform(X_train)

#we extract the pca model now
pca = pca.named_steps['pca']
#lets have a view of the pca variances
print(pd.Series(np.round(pca.explained_variance_ratio_.cumsum(), 4)*100))

In [ ]:
#lets visualise the area under curve graph
# plot feature variance
features = range(pca.n_components_)
cumulative_variance = np.round(np.cumsum(pca.explained_variance_ratio_)*100, decimals=4)
plt.figure(figsize=(12,9)) 
plt.xlabel("No. of components")
plt.ylabel("cumulative variance")
plt.plot(cumulative_variance)

Here __60__ components are enough to describe 98% of the variance in the dataset

In [ ]:
# create pipeline
PCA_VARS = 60
steps = [('scaler', StandardScaler()),
         ("pca", PCA(n_components=PCA_VARS)),
         ("logistic", LogisticRegression(class_weight='balanced'))
        ]
pipeline = Pipeline(steps)

In [ ]:
# fit model
pipeline.fit(X_train, y_train)

# check score on train data
pipeline.score(X_train, y_train)

In [ ]:
#churn prediction over test set
y_pred = pipeline.predict(X_test)

# confusion matrix
conf_mat = confusion_matrix(y_test, y_pred)
print(conf_mat)

#AUC
y_pred_auc = pipeline.predict_proba(X_test)[:, 1]
print("AUC:", round(roc_auc_score(y_test, y_pred_auc),3))

#sensitivity & specificity
sensitivity, specificity, _ = sensitivity_specificity_support(y_test, y_pred, average='binary')
print("sensitivity:", round(sensitivity, 3), "\n", "specificity:", round(specificity, 3), sep='')

### Inculcating Hyperparameter

PCA with Logistic Regression

In [ ]:
#first we calculate the total imbalance figures
y_train.value_counts()/y_train.shape

In [ ]:
pca = PCA()
#cost function
logistic_val = LogisticRegression(class_weight={0:0.1, 1: 0.9})
pca_logistic_val = Pipeline(steps)
steps = [("scaler", StandardScaler()), 
         ("pca", pca),
         ("logistic", logistic_val)
        ]
#Cross validation for creating 5 folds
folds = StratifiedKFold(n_splits = 5, shuffle = True, random_state = 4)
#providing hyperparameters along with logistic penalty l1 & l2
params = {'pca__n_components': [60, 80],'logistic__penalty': ['l1', 'l2'], 'logistic__C': [0.1, 0.5, 1, 2, 3, 4, 5, 10],}
#create gridsearch object
GSCVmodel = GridSearchCV(estimator=pca_logistic_val, cv=folds, scoring='roc_auc', n_jobs=-1,param_grid=params, verbose=1)

In [ ]:
#fitting GSCVmodel
GSCVmodel.fit(X_train, y_train)

In [ ]:
#best possible hyperparameters
print("best possible AUC: ", GSCVmodel.best_score_)
print("best possible hyperparameters: ", GSCVmodel.best_params_)

In [ ]:
#churn prediction over test set
y_pred = GSCVmodel.predict(X_test)

# confusion matrix
conf_mat = confusion_matrix(y_test, y_pred)
print(conf_mat)

#AUC
y_pred_auc = GSCVmodel.predict_proba(X_test)[:, 1]
print("AUC:", round(roc_auc_score(y_test, y_pred_auc),3))

#sensitivity & specificity
sensitivity, specificity, _ = sensitivity_specificity_support(y_test, y_pred, average='binary')
print("sensitivity:", round(sensitivity, 3), "\n", "specificity:", round(specificity, 3), sep='')

In [ ]:
#cross validation 5 folds output
pd.DataFrame(GSCVmodel.cv_results_)

### Random Forest technique

In [ ]:
# random forest with cost function
forest = RandomForestClassifier(class_weight={0:0.1, 1: 0.9}, n_jobs = -1)

##providing hyperparameters
params = {"criterion": ['gini', 'entropy'], "max_features": ['auto', 0.4]}

#Cross validation for creating 5 folds
folds = StratifiedKFold(n_splits = 5, shuffle = True, random_state = 4)

# create gridsearch object
rf_gscvmodel = GridSearchCV(estimator=forest, cv=folds, param_grid=params, scoring='roc_auc', n_jobs=-1, verbose=1)

In [ ]:
#fitting the  RFmodel
rf_gscvmodel.fit(X_train, y_train)

In [ ]:
#best possible hyperparameters
print("best possible AUC: ", rf_gscvmodel.best_score_)
print("best possible hyperparameters: ", rf_gscvmodel.best_params_)

In [ ]:
#churn prediction over test set
y_pred = rf_gscvmodel.predict(X_test)

# confusion matrix
conf_mat = confusion_matrix(y_test, y_pred)
print(conf_mat)

#AUC
y_pred_auc = rf_gscvmodel.predict_proba(X_test)[:, 1]
print("AUC:", round(roc_auc_score(y_test, y_pred_auc),3))

#sensitivity & specificity
sensitivity, specificity, _ = sensitivity_specificity_support(y_test, y_pred, average='binary')
print("sensitivity:", round(sensitivity, 3), "\n", "specificity:", round(specificity, 3), sep='')

We can clearly see that Random Forest in this case is the best model to handle this churn problem

### Important and effective features 

effective feature choice

In [ ]:
#Random forest model to calculate maximum features (IDEA SOURCE: UPGRAD DISCUSSION FORUM)
max_feat = int(round(np.sqrt(X_train.shape[1])))
print(max_feat)

rf_model = RandomForestClassifier(n_estimators=100, class_weight={0:0.1, 1: 0.9}, max_features=max_feat, oob_score=True, random_state=4, verbose=1)


In [ ]:
#fitting rfmodel
rf_model.fit(X_train, y_train)

In [ ]:
#here we display the out of bag error (SOURCE : STACK OVERFLOW)
rf_model.oob_score_

In [ ]:
#churn prediction over test set
y_pred = rf_model.predict(X_test)

# confusion matrix
conf_mat = confusion_matrix(y_test, y_pred)
print(conf_mat)

#AUC
y_pred_auc = rf_model.predict_proba(X_test)[:, 1]
print("AUC:", round(roc_auc_score(y_test, y_pred_auc),3))

#sensitivity & specificity
sensitivity, specificity, _ = sensitivity_specificity_support(y_test, y_pred, average='binary')
print("sensitivity:", round(sensitivity, 3), "\n", "specificity:", round(specificity, 3), sep='')

##### As far as the churn problem is concerned from a business point of view, The Company may focus more on STD, ISD and roaming rates. If these packages are high then it is obvious that most of the users would hesitate. 

##### Also, Internet speed is a major factor in today's world of 3g & 4g. Too high priced packages for considerable less days is also a major backdown which the company must look into.